relay-server/src/utils/rate_limits.rs

```diff
@@ -403,6 +403,67 @@ impl Enforcement {
             })
     }
 
+    /// Applies the [`Enforcement`] on the [`Envelope`] by removing all items that were rate limited
+    /// and emits outcomes for each rate limited category.
+    pub fn apply_with_outcomes(
+        self,
+        envelope: &mut Envelope,
+        scoping: &Scoping,
+        outcome_aggregator: Addr<TrackOutcome>,
```

Comment:
This could take the managed envelope instead and track using the managed envelope.